doc/Memory_Layout.md

```diff
@@ -1,75 +1,112 @@
+address space. While the Cortex-M architecture specifies a high-level layout of
+the address space, the exact layout of Tock can differ from board to board.
+Most boards simply define the beginning and end of flash and SRAM in their
+`layout.ld` file and then include the [generic Tock memory
+map](../boards/kernel_layout.ld).
 
-## Flash
+The kernel has the highest privileges and thus also full control, meaning the
+ability to access all parts of code. Processes on the other hand have specific
+access permissions depending on the memory type.
+
+# Flash
 
 The nonvolatile flash memory hold the kernel code and a linked-list of sorts of
 process code.
 
 ## Kernel code
 
-The kernel code is split into two major regions, `.text` which holds the
-vector table, program code, initialization routines, and other read-only data.
-This section is written to the beginning of flash.
+The kernel code is split into two major regions. The first is `.text`, which
+holds the vector table, program code, initialization routines, and other
+read-only data.  This section is written to the beginning of flash.
+
+The second major region following up the `.text` region is the `.relocate`
+region.  It holds values that need to exist in SRAM, but have non-zero initial
+values that Tock copies from flash to SRAM as part of its initialization (see
+[Startup](Startup.md)).
 
-This is immediately followed by the `.relocate` region, which holds values the
-need to exist in SRAM, but have non-zero initial values that Tock copies from
-flash to SRAM as part of its initialization (see [Startup](Startup.md)).
+The kernel code in flash is not accessible by any process.
 
 ## Process code
 
-Processes are placed in flash starting at a known address which can be retrieved
-in the kernel using the symbol `_sapps`. Each process starts with a Tock Binary
-Format (TBF) header and then the actual application binary. Processes are placed
-continuously in flash, and each process's TBF header includes the entire size of
-the process in flash. This creates a linked-list structure that the kernel uses
-to traverse apps. The end of the valid processes are denoted by an invalid TBF
-header. Typically the flash page after the last valid process is set to all 0x00
-or 0xFF.
+Processes are placed in flash starting at a known address which can be
+retrieved in the kernel using the symbol `_sapps`. Each process starts with a
+Tock Binary Format (TBF) header and then the actual application binary.
+Processes are placed continuously in flash, and each process's TBF header
+includes the entire size of the process in flash. This creates a linked-list
+structure that the kernel uses to traverse apps. The end of the valid processes
+are denoted by an invalid TBF header. Typically the flash page after the last
+valid process is set to all 0x00 or 0xFF.
+
+Processes have read-only access to their own memory in flash, but can not read
+other processes' code. 
```

Comment:
Processes can, however, write their own flash using a system call (if the board supports it).

Reply:
Are there limits to this? i.e. can't write their own header (or even their own code?). I thought we had a dedicated 'app flash' area they wrote to? I don't totally remember how this works, but this would probably be a good place to document it

Reply:
Correct, they cannot overwrite the protected region of the app (as defined by the kernel) which includes the app header. However, from the perspective of this document, I think it is fair to assume that apps can write to any other part of their own flash (assuming the board supports it). The functions in the kernel crate that the app_flash capsule uses only protect the app header.